File: 백준/Silver/18232. 텔레포트 정거장/텔레포트 정거장.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	
    	adj = new ArrayList[N+1];
    	for(int i = 1; i<=N; i++) {
    		adj[i] = new ArrayList<>();
    	}
    	
    	st = new StringTokenizer(br.readLine());
    	S = Integer.parseInt(st.nextToken());
    	E = Integer.parseInt(st.nextToken());
    	
    	for(int i = 0; i < M; i++) {
    		st = new StringTokenizer(br.readLine());
    		int x = Integer.parseInt(st.nextToken());
    		int y = Integer.parseInt(st.nextToken());
    		adj[x].add(y);
    		adj[y].add(x);
    	}
    	
    	dist = new int[N+1];
    	Arrays.fill(dist, -1);
    	
    	Queue<Integer> q = new ArrayDeque<>();
    	dist[S] = 0;
    	q.add(S);
    	
    	while(!q.isEmpty()) {
    		int cur = q.poll();
    		
    		for(int dir : new int[] {-1,1}) {
    			int nxt = cur + dir;
    			if(OOB(nxt) || dist[nxt] != -1)
    				continue;
    			
    			dist[nxt] = dist[cur] + 1;
    			q.add(nxt);
    		}
    		
    		for(int nxt : adj[cur]) {
    			if(OOB(nxt) || dist[nxt] != -1)
    				continue;
    			dist[nxt] = dist[cur] + 1;
    			q.add(nxt);
    		}
    	}
    	
    	System.out.println(dist[E]);
    	
    }
    
    static boolean OOB(int x) {
    	return x<=0 || x> N;
    }
    static StringTokenizer st;
    static int N,M, S, E;
    static ArrayList<Integer> adj[];
    static int dist[];
}

File: 백준/Silver/18232. 텔레포트 정거장/텔레포트 정거장.py
from collections import deque

N,M = map(int, input().split())

adj = [[] for _ in range(N+1)]

S, E = map(int,input().split())

for i in range(M):
    x, y = map(int, input().split())

    adj[x].append(y)
    adj[y].append(x)
dist = [-1 for _ in range(N+1)]

q = deque()
dist[S] = 0
q.append(S)
def OOB(x : int):
    return x<= 0 or x> N

while q:
    # print(f"q: {q}, dist: {dist}")
    cur = q.popleft()

    for dir in [-1,1]:
        nxt = cur + dir
        if OOB(nxt) or dist[nxt] != -1:
            continue

        dist[nxt] = dist[cur] + 1
        q.append(nxt)

    for nxt in adj[cur]:
        if OOB(nxt) or dist[nxt] != -1:
            continue

        dist[nxt] = dist[cur] + 1
        q.append(nxt)
print(dist[E])